sender_pool.go

```diff
@@ -119,62 +136,97 @@ func WithMaxSenders(count int) LineSenderPoolOption {
-func (p *LineSenderPool) Acquire(ctx context.Context) (LineSender, error) {
+// If there is already maximum number of senders obtained from the pool,
+// this call will block until one of the senders is returned back to
+// the pool by calling sender.Close().
+func (p *LineSenderPool) Sender(ctx context.Context) (LineSender, error) {
+	var (
+		s   LineSender
+		err error
+	)
+
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
 	if p.closed {
-		return nil, fmt.Errorf("cannot Acquire a LineSender from a closed LineSenderPool")
+		return nil, errAcquireFromClosedPool
+	}
+
+	// We may have to wait for a free sender
+	for len(p.freeSenders) == 0 && p.numSenders == p.maxSenders {
+		p.cond.Wait()
 	}
 
-	if len(p.senders) > 0 {
+	if p.closed {
+		return nil, errAcquireFromClosedPool
+	}
+
+	if len(p.freeSenders) > 0 {
 		// Pop sender off the slice and return it
-		s := p.senders[len(p.senders)-1]
-		p.senders = p.senders[0 : len(p.senders)-1]
+		s := p.freeSenders[len(p.freeSenders)-1]
+		atomic.AddUint64(&s.tick, 1)
+		p.freeSenders = p.freeSenders[0 : len(p.freeSenders)-1]
 		return s, nil
 	}
 
 	if p.conf != "" {
-		return LineSenderFromConf(ctx, p.conf)
+		s, err = LineSenderFromConf(ctx, p.conf)
 	} else {
 		conf := newLineSenderConfig(httpSenderType)
 		for _, opt := range p.opts {
 			opt(conf)
 			if conf.senderType == tcpSenderType {
-				return nil, errors.New("tcp/s not supported for pooled senders, use http/s only")
+				return nil, errHttpOnlySender
 			}
 		}
-		return newHttpLineSender(conf)
+		s, err = newHttpLineSender(conf)
 	}
 
+	if err != nil {
+		return nil, err
+	}
+
+	p.numSenders++
+
+	ps := &pooledSender{
+		pool:    p,
+		wrapped: s,
+	}
+	atomic.AddUint64(&ps.tick, 1)
+	return ps, nil
 }
 
-// Release flushes the LineSender and returns it back to the pool. If the pool
-// is full, the sender is closed and discarded. In cases where the sender's
-// flush fails, it is not added back to the pool.
-func (p *LineSenderPool) Release(ctx context.Context, s LineSender) error {
-	// If there is an error on flush, do not add the sender back to the pool
-	if err := s.Flush(ctx); err != nil {
-		return err
+func (p *LineSenderPool) free(ctx context.Context, ps *pooledSender) error {
+	var flushErr error
+
+	if !ps.dirty {
+		flushErr = ps.Flush(ctx)
 	}
 
 	p.mu.Lock()
 	defer p.mu.Unlock()
-
-	for i := range p.senders {
-		if p.senders[i] == s {
-			return fmt.Errorf("LineSender %p has already been released back to the pool", s)
+	// Notify free sender waiters, if any
+	defer p.cond.Broadcast()
+
+	if flushErr != nil {
+		// Failed to flush, close and call it a day
+		p.numSenders--
+		closeErr := ps.wrapped.Close(ctx)
+		if closeErr != nil {
+			return fmt.Errorf("%s %w", flushErr, closeErr)
 		}
+		return flushErr
 	}
 
-	if p.closed || len(p.senders) >= p.maxSenders {
-		return s.Close(ctx)
+	if ps.dirty || p.closed {
+		// Previous error or closed pool, close and call it a day
+		p.numSenders--
+		return ps.wrapped.Close(ctx)
```

Comment:
How can we cleanly close a dirty sender? Won't we attempt to flush again on this close, which could be problematic if the sender is in an errored state to begin with?

Reply:
`Close` call is a must even if the previous `Flush` call errored. Even if we attempt to flush, that's fine here since it's documented that `Close` on a pooled sender may call `Flush`.